### ISSUE TITLE
MaxPool3d makes model not reproducible.

### ISSUE THREAD

**Issue by yelantingfeng** [pytorch/pytorch#23550]
## 🐛 Bug

MaxPool3d makes gradient not deterministic.

## To Reproduce

1. Run following code with command `python test1.py`. It saves model weights, intermediate variables and gradients into .pth files.
```python
# test1.py
import torch
import torch.nn as nn
import random

torch.manual_seed(1)
random.seed(1)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

class model(nn.Module):
    def __init__(self):
        super(model, self).__init__()
        self.conv = nn.Conv3d(3,64,(1,7,7),stride=(1,2,2),padding=(0,3,3),bias=False)
        self.relu = nn.ReLU(inplace=True)
        self.maxpool = nn.MaxPool3d((1,3,3),stride=(1,2,2))
    def forward(self, x):
        return self.maxpool(self.relu(self.conv(x)))
        #return self.relu(self.conv(x))

a = torch.rand(2,3,8,64,64)
m = model()
a = a.to("cuda:0")
m.to("cuda:0")
b = m(a)
b = b.view(2, -1)
fc = nn.Linear(b.size(1), 80)
fc.to("cuda:0")
pred = fc(b)
label = torch.randint(2, (2,80), dtype=torch.float32, device="cuda:0")
bceloss = nn.BCEWithLogitsLoss()
loss = bceloss(pred, label)
print(loss.item())

loss.backward()
tensor_dict = dict(a=a,b=b,pred=pred,label=label,loss=loss)
torch.save(tensor_dict, "tensor1.pth")
model_dict = dict(conv_w=m.conv.weight, fc_w=fc.weight, fc_b=fc.bias)
torch.save(model_dict, "model1.pth")
grad_dict = dict(conv_w_grad=m.conv.weight.grad, fc_w_grad = fc.weight.grad, fc_b_grad = fc.bias.grad)
torch.save(grad_dict, "grad1.pth")
```
2. Run following script with command `python test2.py`. It does the same things, but save targets into different files.
```python
# test2.py
import torch
import torch.nn as nn
import random

torch.manual_seed(1)
random.seed(1)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

class model(nn.Module):
    def __init__(self):
        super(model, self).__init__()
        self.conv = nn.Conv3d(3,64,(1,7,7),stride=(1,2,2),padding=(0,3,3),bias=False)
        self.relu = nn.ReLU(inplace=True)
        self.maxpool = nn.MaxPool3d((1,3,3),stride=(1,2,2))
    def forward(self, x):
        return self.maxpool(self.relu(self.conv(x)))
        #return self.relu(self.conv(x))

a = torch.rand(2,3,8,64,64)
m = model()
a = a.to("cuda:0")
m.to("cuda:0")
b = m(a)
b = b.view(2, -1)
fc = nn.Linear(b.size(1), 80)
fc.to("cuda:0")
pred = fc(b)
label = torch.randint(2, (2,80), dtype=torch.float32, device="cuda:0")
bceloss = nn.BCEWithLogitsLoss()
loss = bceloss(pred, label)
print(loss.item())

loss.backward()
tensor_dict = dict(a=a,b=b,pred=pred,label=label,loss=loss)
torch.save(tensor_dict, "tensor2.pth")
model_dict = dict(conv_w=m.conv.weight, fc_w=fc.weight, fc_b=fc.bias)
torch.save(model_dict, "model2.pth")
grad_dict = dict(conv_w_grad=m.conv.weight.grad, fc_w_grad = fc.weight.grad, fc_b_grad = fc.bias.grad)
torch.save(grad_dict, "grad2.pth")
```
3. Run following script by using `python cmp.py` to compare those saved files.
```python
# cmp.py
import torch

def relative_diff(a,b):
    rel = abs(a-b)/abs(a)
    zero_mask = (a==0)&(b==0)
    rel[zero_mask] = 0
    rel[torch.isnan(rel)]=1000
    return rel

def cmp(path1, path2):
    cmp1 = torch.load(path1, map_location="cpu")
    cmp2 = torch.load(path2, map_location="cpu")
    for key in cmp1:
        cmp11, cmp22 = cmp1[key], cmp2[key]
        diff = relative_diff(cmp11, cmp22)
        print("max of {} diff: {}".format(key, diff.max().item()))
        print("sum of {} diff: {}".format(key, diff.sum().item()))

tensor1 = "tensor1.pth"
tensor2 = "tensor2.pth"
print("Compare between tensors:")
cmp(tensor1, tensor2)

model1 = "model1.pth"
model2 = "model2.pth"
print("Compare between model weights:")
cmp(model1, model2)

grad1 = "grad1.pth"
grad2 = "grad2.pth"
print("Compare between gradient:")
cmp(grad1, grad2)
```

## Expected behavior
After running the compare script `cmp.py`, I got following outputs.
```
Compare between tensors:
max of a diff: 0.0
sum of a diff: 0.0
max of b diff: 0.0
sum of b diff: 0.0
max of pred diff: 0.0
sum of pred diff: 0.0
max of label diff: 0.0
sum of label diff: 0.0
max of loss diff: 0.0
sum of loss diff: 0.0
Compare between model weights:
max of conv_w diff: 0.0
sum of conv_w diff: 0.0
max of fc_w diff: 0.0
sum of fc_w diff: 0.0
max of fc_b diff: 0.0
sum of fc_b diff: 0.0
Compare between gradient:
max of conv_w_grad diff: 0.03999999910593033
sum of conv_w_grad diff: 0.04221801832318306
max of fc_b_grad diff: 0.0
sum of fc_b_grad diff: 0.0
max of fc_w_grad diff: 0.0
sum of fc_w_grad diff: 0.0
```
It shows that the gradient of the weight in convolution layer is not deterministic, but after removing that MaxPool3d layer from the model, all differences become zero. When use this in a practical model, as training goes on, the difference of gradient leads to totally different models. 

I do know that some atomic operations may cause the result unreproducible, but does MaxPool3d use that kind of operations? This difference seems too much, and I wonder if there is a workaround to avoid this. Any help you can provide will be appreciated.

## Environment

```
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.5
Is CUDA available: Yes
CUDA runtime version: 9.2.148
GPU models and configuration: 
GPU 0: Tesla V100-PCIE-16GB
GPU 1: Tesla V100-PCIE-16GB
GPU 2: Tesla V100-PCIE-16GB
GPU 3: Tesla V100-PCIE-16GB
GPU 4: Tesla V100-PCIE-16GB
GPU 5: Tesla V100-PCIE-16GB
GPU 6: Tesla V100-PCIE-16GB
GPU 7: Tesla V100-PCIE-16GB

Nvidia driver version: 396.37
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.1

Versions of relevant libraries:
[pip3] numpy==1.16.4
[pip3] torch==1.1.0
[conda] Could not collect
```

**Comment by albanD** [comment]
Closing this in favor of https://github.com/pytorch/pytorch/issues/15359 that aims at adding a deterministic mode to solve this.
